drv/riscv-plic-server/src/main.rs

```diff
@@ -0,0 +1,202 @@
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                for irq in irqs.iter() {
+                    plic.mask(0, irq.0.try_into().unwrap());
+                }
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(ExtIntCtrlError::IRQUnassigned)),
+        }
+    }
+
+    /// Enables the selected interrupt on the PLIC.
+    fn enable_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<ExtIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                for irq in irqs.iter() {
+                    plic.unmask(0, irq.0.try_into().unwrap());
+                }
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(ExtIntCtrlError::IRQUnassigned)),
+        }
+    }
+
+    /// Completes the interrupt on the PLIC, allowing for a new one to come
+    /// through.
+    fn complete_int(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        irq: u32,
+    ) -> Result<(), RequestError<ExtIntCtrlError>> {
+        let owner: InterruptOwner = InterruptOwner {
+            task: msg.sender.index() as u32,
+            notification: irq,
+        };
+
+        match get_task_irqs(owner) {
+            Ok(irqs) => {
+                let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+                for irq in irqs.iter() {
+                    plic.complete(0, irq.0.try_into().unwrap());
+                }
+
+                return Ok(());
+            }
+            Err(()) => return Err(Runtime(ExtIntCtrlError::IRQUnassigned)),
+        }
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl {
+    // The PLIC is only interested in interrupt notifications from the kernel
+    fn current_notification_mask(&self) -> u32 {
+        return PLIC_IRQ;
+    }
+
+    // An interrupt has come in.
+    // NOTE: Currently, the driver assumes that all interrupts come in on
+    //       Context 0.
+    fn handle_notification(&mut self, _bits: u32) {
+        let plic = unsafe { &mut *PLIC_REGISTER_BLOCK };
+        loop {
+            let irq: u32 = match plic.claim(0) {
+                Some(irq) => core::primitive::u16::from(irq) as u32,
+                None => break,
+            };
+
+            // An error means an interrupt came in on a line that no task has
+            // ownership over.
+            let owner: (TaskId, u32) = match get_irq_owner(irq) {
+                Ok(owner) => owner,
+                Err(()) => continue,
+            };
+
+            let code = sys_post(owner.0, owner.1);
+
+            // The task that owns the line was restarted.
```

Comment:
Should we disable a irq if the task that owns it was restarted?  
A task may have startup code that expects interrupts not yet be enabled.  How does the arm port handle this?

Reply:
There's not any straightforward way to add this. We would have to somehow inform the interrupt controller that a task failed, which isn't possible unless the PLIC has a notification for said task, but at that point it's possible the task has already been restarted and is ready to receive interrupts again.

I don't think this is something we have to worry about, a task can just tell the controller to disable its' interrupts at the start of main().

Reply:
I think since this is a `sys_post` operation, it will only set the notification bits. So, unless the task asks the kernel for notifications, it won't really be interrupted.

Reply:
Also, as far as I know the PLIC spec doesn't say that disabling an interrupt will actually disable it on the device, so this would require a lot more work than just figuring out `disable_int()` on a task's death.